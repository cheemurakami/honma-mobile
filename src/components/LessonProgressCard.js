import React from "react";
import styled from "styled-components/native";

const LessonProgressCard = ({ dialect, grammars }) => {
  const availableGrammars = grammars
    ? grammars.filter(
        (grammar) =>
          grammar.dialect_id == dialect.id && grammar.quizzes.length > 0
      )
    : [];

  const lessonsCount = availableGrammars.length;

  const quizzes = availableGrammars
    .map((grammar) => {
      return grammar.quizzes;
    })
    .flat();

  const completedLessonsCount = () => {
    let grammarIncompletedList = {};

    if (quizzes.length > 0) {
      quizzes.forEach((quiz) => {
        if (quiz) {
          if (quiz.quiz_completed === null) {
            grammarIncompletedList[quiz.grammar_id] = true;
          }
        }
      });
    }
    const imcompletedLessonsCount = Object.keys(grammarIncompletedList).length;
    return lessonsCount - imcompletedLessonsCount;
  };

  return (
    <ProgressIcon>
      <ProgressText>
        {completedLessonsCount()} / {lessonsCount}
      </ProgressText>
      <ProgressText>Lessons</ProgressText>
    </ProgressIcon>
  );
};

const ProgressIcon = styled.View`
  width: 70px;
  height: 45px;
  margin: 10px;
  border-radius: 10px;
  background-color: #f6a704;
  align-items: center;
  justify-content: center;
`;
const ProgressText = styled.Text`
  color: #fff;
  font-weight: bold;
`;

export default LessonProgressCard;
