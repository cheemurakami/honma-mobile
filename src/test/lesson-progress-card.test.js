import React from "react";
import renderer from "react-test-renderer";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import rootReducer from "../rdx/reducers/index.js";
import LessonProgressCard from "../components/LessonProgressCard.js";
import { waitFor, render, screen } from "@testing-library/react-native";
import { dialectData, grammarsData } from "./test-data/lesson-test-data.js";

describe("LessonProgressCard tests", () => {
  const store = configureStore({ reducer: rootReducer });
  const renderComponent = () => {
    return render(
      <Provider store={store}>
        <LessonProgressCard dialect={dialectData} grammars={grammarsData} />
      </Provider>
    );
  };

  it("matches snapshot", async () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <LessonProgressCard dialect={dialectData} grammars={grammarsData} />
        </Provider>
      )
      .toJSON();
    await waitFor(() => {
      expect(tree).toMatchSnapshot();
    });
  });

  it("renders completed lesson number and available lessons number", async () => {
    renderComponent();

    const cardText = screen.getByText("1 / 2");

    await waitFor(() => {
      expect(cardText).toBeTruthy();
    });
  });
});
