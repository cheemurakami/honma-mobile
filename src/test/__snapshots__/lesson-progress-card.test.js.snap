// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonProgressCard tests matches snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#f6a704",
        "borderRadius": 10,
        "height": 45,
        "justifyContent": "center",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 10,
        "width": 70,
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#fff",
          "fontWeight": "bold",
        },
      ]
    }
  >
    1
     / 
    2
  </Text>
  <Text
    style={
      [
        {
          "color": "#fff",
          "fontWeight": "bold",
        },
      ]
    }
  >
    Lessons
  </Text>
</View>
`;
